From 24583927d4d6a9b8ebeb0b71f8896f7f14e45206 Mon Sep 17 00:00:00 2001
From: Michael Buesch <mb@bu3sch.de>
Date: Wed, 9 Mar 2011 18:07:05 +0000
Subject: [PATCH 0320/1092] cbus-tahvo-usb: Fix tahvo_usb_device pointer

The static tahvo_usb_device is uninitialized, but used
in the otg interrupt handler. This results in a NULL pointer
dereference on interrupt.
Fix this by storing a struct tahvo_usb pointer instead of
a platform device.

Signed-off-by: Michael Buesch <mb@bu3sch.de>
Acked-by: Felipe Balbi <balbi@ti.com>
Signed-off-by: Tony Lindgren <tony@atomide.com>
---
 drivers/cbus/tahvo-usb.c | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/drivers/cbus/tahvo-usb.c b/drivers/cbus/tahvo-usb.c
index 3b83bab..1ada7d3 100644
--- a/drivers/cbus/tahvo-usb.c
+++ b/drivers/cbus/tahvo-usb.c
@@ -99,7 +99,7 @@ struct tahvo_usb {
 	int tahvo_mode;
 #endif
 };
-static struct platform_device tahvo_usb_device;
+static struct tahvo_usb *tahvo_usb_device;
 
 /*
  * ---------------------------------------------------------------------------
@@ -114,8 +114,7 @@ static struct platform_device *tahvo_otg_dev;
 
 static irqreturn_t omap_otg_irq(int irq, void *arg)
 {
-	struct platform_device *otg_dev = arg;
-	struct tahvo_usb *tu = platform_get_drvdata(otg_dev);
+	struct tahvo_usb *tu = arg;
 	u16 otg_irq;
 
 	otg_irq = omap_readw(OTG_IRQ_SRC);
@@ -201,12 +200,12 @@ static int __init omap_otg_probe(struct platform_device *pdev)
 
 	return request_irq(tahvo_otg_dev->resource[1].start,
 			   omap_otg_irq, IRQF_DISABLED, DRIVER_NAME,
-			   &tahvo_usb_device);
+			   tahvo_usb_device);
 }
 
 static int __exit omap_otg_remove(struct platform_device *pdev)
 {
-	free_irq(tahvo_otg_dev->resource[1].start, &tahvo_usb_device);
+	free_irq(tahvo_otg_dev->resource[1].start, tahvo_usb_device);
 	tahvo_otg_dev = NULL;
 
 	return 0;
@@ -659,6 +658,7 @@ static int __init tahvo_usb_probe(struct platform_device *pdev)
 	tu = kzalloc(sizeof(*tu), GFP_KERNEL);
 	if (!tu)
 		return -ENOMEM;
+	tahvo_usb_device = tu;
 
 	tu->pt_dev = container_of(dev, struct platform_device, dev);
 #ifdef CONFIG_USB_OTG
@@ -682,6 +682,7 @@ static int __init tahvo_usb_probe(struct platform_device *pdev)
 				(unsigned long) tu, "vbus_interrupt");
 	if (ret != 0) {
 		kfree(tu);
+		tahvo_usb_device = NULL;
 		printk(KERN_ERR "Could not register Tahvo interrupt for VBUS\n");
 		return ret;
 	}
@@ -708,6 +709,7 @@ static int __init tahvo_usb_probe(struct platform_device *pdev)
 	ret = otg_set_transceiver(&tu->otg);
 	if (ret < 0) {
 		printk(KERN_ERR "Cannot register USB transceiver\n");
+		tahvo_usb_device = NULL;
 		kfree(tu);
 		tahvo_free_irq(TAHVO_INT_VBUSON);
 		return ret;
@@ -732,6 +734,8 @@ static int __exit tahvo_usb_remove(struct platform_device *pdev)
 #ifdef CONFIG_USB_OTG
 	device_remove_file(&pdev->dev, &dev_attr_otg_mode);
 #endif
+	tahvo_usb_device = NULL;
+
 	return 0;
 }
 
-- 
1.7.11.2

